### PR TITLE
csv module utils: detect unicode BOM in content

### LINKED ISSUE #544: read_csv - Key 'Name' was not found in the CSV header fields
##### SUMMARY
The `read_csv` module fails to identify a field, yet displaces the field in the list of available fields.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
read_csv

##### ANSIBLE VERSION

```
ansible 2.9.10
  config file = /home/anton/git/ansible-deploy-vmware-vm/ansible.cfg
  configured module search path = ['/home/anton/git/ansible-deploy-vmware-vm/library']
  ansible python module location = /home/anton/.local/lib/python3.6/site-packages/ansible
  executable location = /home/anton/.local/bin/ansible
  python version = 3.6.9 (default, Apr 18 2020, 01:56:04) [GCC 8.4.0]

```

##### CONFIGURATION

```
# config file for ansible -- http://ansible.com/
# ==============================================

# nearly all parameters can be overridden in ansible-playbook
# or with command line flags. ansible will read ANSIBLE_CONFIG,
# ansible.cfg in the current working directory, .ansible.cfg in
# the home directory or /etc/ansible/ansible.cfg, whichever it
# finds first

[defaults]
host_key_checking = False
host_key_check = False
ansible_python_interpreter=/usr/bin/python3
log_path = ./ansible.log
#bin_ansible_callbacks=True
#stdout_callback = debug


# some basic default values...
library        = ./library

# additional paths to search for roles in, colon separated
roles_path    = ./roles

[ssh_connection]
# ssh arguments to use
ssh_args = -o StrictHostKeyChecking=no
timeout=60


```

##### OS / ENVIRONMENT
Ubuntu 20:04

##### STEPS TO REPRODUCE

```
---
- name: Right-size VMs
  gather_facts: false
  hosts: all
  connection: local
  tasks:
    # Read a CSV file and access the first item
  - name: Read users from CSV file and return a list
    read_csv:
        path: "files/vms/6-19-20 Optimization Report - Oversized Virtual Machines Prod2.csv"
        key: Name
    register: users

  - debug:
      msg: 'User {{ users.list.2.Name}}'
        # msg: 'User {{ users.list.2.Name}} has UID {{ users.list.2.ReclaimablevCPUs}} and GID {{ users.list.2.ReclaimableMemory}}'
        # msg: "{{ users }}"



```

##### EXPECTED RESULTS
Expect to be able to read CSV values by col name (field) as based on module documentation.

##### ACTUAL RESULTS
```
fatal: [localhost]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python"}, "changed": false, "msg": "Key 'Name' was not found in the CSV header fields: ﻿Name, Configured-vCPU, ReclaimablevCPUs, ConfiguredMemory, ReclaimableMemory, ParentvCenter"}
```

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY

If CSV content contains a BOM marker then remove it from the content

Fixes #544 

##### ISSUE TYPE

- Bugfix Pull Request
- Test Pull Request

##### COMPONENT NAME

plugins/module_utils/csv.py

### CODE REVIEW

File: changelogs/fragments/6662-csv-bom.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - csv module utils - detects and remove unicode BOM markers from incoming CSV content ().
```
Comment (felixfontein) [inline changelogs/fragments/6662-csv-bom.yml:2]: Something missing here ;-)

File: plugins/module_utils/csv.py [hunk 2]
```diff
@@ -55,12 +55,17 @@ class unix_dialect(csv.Dialect):
...
 
     if PY3:
+        if data.startswith(BOM):
+            data = data[1:]
         fake_fh = StringIO(data)
     else:
+        # type(data) == bytes in Py2, and encoded BOM is 3 bytes long
+        if data.startswith(BOM.encode('utf-8')):
+            data = data[3:]
         fake_fh = BytesIO(data)
```
Comment (felixfontein) [inline plugins/module_utils/csv.py:69]: How about doing `BOM = to_native(u'\ufeff')` and then `if data.startswith(BOM): data = data[len(BOM):]` before the `if PY3`? That's easier to understand IMO and it avoids magic numbers.

Comment (russoz) [reply plugins/module_utils/csv.py]: much better indeed

File: changelogs/fragments/6662-csv-bom.yml (hunk 1 above)
Comment (russoz) [reply changelogs/fragments/6662-csv-bom.yml]: done

**Comment by felixfontein** [comment]
recheck

**Review by felixfontein — APPROVED** [review]
Looks good to me. If nobody objects, I'll merge this in ~a week.

**Comment by felixfontein** [comment]
@russoz thanks for fixing this!

**Pull request merged**